lib/ecto/association.ex

```diff
@@ -264,8 +267,12 @@ defmodule Ecto.Association.Has do
   end
 
   @doc false
-  def build(%{assoc: assoc, owner_key: owner_key, assoc_key: assoc_key}, struct) do
-    Map.put apply(assoc, :__struct__, []), assoc_key, Map.get(struct, owner_key)
+  def build(%{assoc: assoc, owner_key: owner_key, assoc_key: assoc_key,
+              queryable: queryable}, struct) do
+    assoc
+    |> apply(:__struct__, [])
+    |> Map.put(assoc_key, Map.get(struct, owner_key))
+    |> Ecto.Schema.put_meta(:source, Ecto.Association.source_from_query(queryable))
```

Comment:
We should only put the source if there is one from the association. Otherwise it will be set to nil and that is not good. We should write a test for that. :)